### PR TITLE
[WIP] Store relative paths in path mapping repositories

### PR DESCRIPTION
**Pull request by tgalopin**
Fix https://github.com/puli/issues/issues/117.

This PR aims to implement a base directory for `PathMappingRepository` and `OptimizedPathMappingRepository` in order to store relative paths in the underlying key-value stores.

Here is the plan:
1. Put all the fixtures (even the temporary ones) in the `tests/Fixtures` directory **done**
2. Add testing for base directory check and stores **done**
3. Implement the feature to fix the tests **done**

### CODE REVIEW

**Comment by tgalopin** [comment]
This PR seems to be working properly. Can you review @webmozart?

**Comment by tgalopin** [comment]
Well, I spoke too fast, it does not work on Windows. I'll work on that.

File: src/AbstractPathMappingRepository.php [hunk 1]
```diff
@@ -278,4 +299,54 @@ protected function createVirtualResource($path = null)
...
+    /**
+     * Transform a collection of relative paths into a collection of absolute paths.
+     *
+     * @param array $relativePaths
+     *
+     * @return array
+     */
+    protected function makePathsCollectionAbsolute($relativePaths)
+    {
+        foreach ($relativePaths as &$relativePath) {
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: I'd rather include the index in the loop and then assign by index rather than by reference. Using references in loops can cause bugs that are hard to spot.

File: src/AbstractPathMappingRepository.php [hunk 2]
```diff
@@ -278,4 +299,54 @@ protected function createVirtualResource($path = null)
...
+    }
+
+    /**
+     * Transform a collection of relative paths into a collection of absolute paths.
+     *
+     * @param array $relativePaths
+     *
+     * @return array
+     */
+    protected function makePathsCollectionAbsolute($relativePaths)
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: `makePathsAbsolute()` should suffice

File: src/AbstractPathMappingRepository.php [hunk 3]
```diff
@@ -278,4 +299,54 @@ protected function createVirtualResource($path = null)
...
+    /**
+     * Transform a relative path into an absolute path.
+     *
+     * @param string $absolutePath
+     *
+     * @return string
+     */
+    protected function makePathRelative($absolutePath)
+    {
+        if (0 === strpos($absolutePath, 'l:')) {
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: Why can this be called with a link?

File: src/AbstractPathMappingRepository.php [hunk 4]
```diff
@@ -205,8 +222,12 @@ protected function createResource($filesystemPath, $path = null)
             return $this->createLinkResource(substr($filesystemPath, 2), $path);
         }
 
-        if ($filesystemPath && file_exists($filesystemPath)) {
-            return $this->createFilesystemResource($filesystemPath, $path);
+        if ($filesystemPath) {
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: Can you make this check more concrete?

File: src/AbstractPathMappingRepository.php [hunk 5]
```diff
@@ -101,6 +108,16 @@ private function addResource($path, $resource)
         if ($resource instanceof LinkResource) {
             $this->addLinkResource($path, $resource);
         } else {
+            // Check resource is in base directory
+            if (!Path::isBasePath($this->baseDirectory, $resource->getFilesystemPath())) {
+                throw new UnsupportedResourceException(sprintf(
+                    'This PathMapping repository is based on directory "%s" and '.
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: This path-mapping repository...

File: src/AbstractPathMappingRepository.php (hunk 3 above)
Comment (tgalopin) [inline src/AbstractPathMappingRepository.php]: Because under Windows, `l:/directory` will be understood as a partition where it shouldn't. It avoids problems.

Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: I mean, why is this method called at all when adding links?

File: src/AbstractPathMappingRepository.php (hunk 1 above)
Comment (tgalopin) [inline src/AbstractPathMappingRepository.php]: Good idea :) .

File: src/AbstractPathMappingRepository.php (hunk 3 above)
Comment (tgalopin) [inline src/AbstractPathMappingRepository.php]: It isn't :D . I change that.

File: src/AbstractPathMappingRepository.php [hunk 6]
```diff
@@ -43,13 +43,20 @@
...
     /**
+     * @var string
+     */
+    protected $baseDirectory;
+
+    /**
      * Creates a new repository.
      *
-     * @param KeyValueStore $store The store of all the paths.
+     * @param KeyValueStore $store
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: You removed the description here by accident.

File: src/AbstractPathMappingRepository.php [hunk 7]
```diff
@@ -101,6 +108,16 @@ private function addResource($path, $resource)
         if ($resource instanceof LinkResource) {
             $this->addLinkResource($path, $resource);
         } else {
+            // Check resource is in base directory
+            if (!Path::isBasePath($this->baseDirectory, $resource->getFilesystemPath())) {
+                throw new UnsupportedResourceException(sprintf(
+                    'This path-mapping repository is based on directory "%s" and '.
+                    'therefore does not support filesystem resource "%s"',
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: I would reword this exception message to be more easily understandable:

> Can only add resources in %s. Tried to add %s.

File: src/AbstractPathMappingRepository.php [hunk 8]
```diff
@@ -278,4 +301,49 @@ protected function createVirtualResource($path = null)
...
+    /**
+     * Transform a relative path into an absolute path.
+     *
+     * @param string $relativePath
+     *
+     * @return string
+     */
+    protected function makePathAbsolute($relativePath)
+    {
+        if (0 === strpos($relativePath, 'l:')) {
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: Is this method ever called with a link?

Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: Ah I see it can. I would rename this method (and the plural version before) to `resolveRelativePath()`.

File: tests/OptimizedPathMappingRepositoryTest.php [hunk 9]
```diff
@@ -170,4 +105,33 @@ public function testRemoveFailsIfLanguageNotGlob()
...
+    {
+        $repo = $this->createBaseDirectoryRepository($this->store, __DIR__.'/Fixtures');
+        $repo->add('/webmozart/file', new FileResource(__DIR__.'/Fixtures/dir1/file1'));
+        $repo->add('/webmozart/dir', new DirectoryResource(__DIR__.'/Fixtures/dir2'));
+
+        $this->assertTrue($repo->contains('/webmozart/file'));
+        $this->assertTrue($repo->contains('/webmozart/dir'));
+        $this->assertEquals('dir1/file1', $this->store->get('/webmozart/file'));
+        $this->assertEquals('dir2', $this->store->get('/webmozart/dir'));
+        $this->assertPathsEquals(__DIR__.'/Fixtures/dir1/file1', $repo->get('/webmozart/file')->getFilesystemPath());
```
Comment (webmozart) [inline tests/OptimizedPathMappingRepositoryTest.php]: `getFilesystemPath()` should already return a normalized path. Hence it should be sufficient to do:

```php
$this->assertSame(Path::normalize(__DIR__.'/Fixtures/dir1/file1'), $repo->get('/webmozart/file')->getFilesystemPath());
```

This is also the way I solved failing Windows tests in other repositories.

You can remove `assertPathsEquals()`.

Comment (webmozart) [inline tests/OptimizedPathMappingRepositoryTest.php]: It's probably even easier to add a property `$fixturesDir` to the test and set it in `setUp()`:

```php
$this->fixturesDir = Path::normalize(__DIR__.'/Fixtures');
```

Then you don't need to call `normalize()` in the individual tests.

File: tests/PathMappingRepositoryTest.php [hunk 10]
```diff
@@ -17,71 +17,30 @@
...
-
-        $filesystem = new Filesystem();
-        $filesystem->remove($this->tempDir);
+        return new PathMappingRepository($store, $baseDirectory);
     }
 
     protected function createPrefilledRepository(Resource $root)
     {
-        $repo = new PathMappingRepository(new ArrayStore());
+        $repo = new PathMappingRepository(new ArrayStore(), ('\\' === DIRECTORY_SEPARATOR) ? 'C:/' : '/');
```
Comment (webmozart) [inline tests/PathMappingRepositoryTest.php]: Can we guarantee that this is always `C:/`? This smells like a volatile test to me.

**Comment by tgalopin** [comment]
All green \o/ (too rare to not notice :D ).

File: src/AbstractPathMappingRepository.php [hunk 11]
```diff
@@ -101,6 +108,15 @@ private function addResource($path, $resource)
         if ($resource instanceof LinkResource) {
             $this->addLinkResource($path, $resource);
         } else {
+            // Check resource is in base directory
+            if (!Path::isBasePath($this->baseDirectory, $resource->getFilesystemPath())) {
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: You can coalesce this into an `if - elseif - else`

File: src/AbstractPathMappingRepository.php [hunk 12]
```diff
@@ -278,4 +300,49 @@ protected function createVirtualResource($path = null)
...
+    /**
+     * Transform a relative path into an absolute path.
+     *
+     * @param string $absolutePath
+     *
+     * @return string
+     */
+    protected function makePathRelative($absolutePath)
+    {
+        return Path::makeRelative($absolutePath, $this->baseDirectory);
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: Let's remove this method and use `Path::makeRelative()` directly. Less method calls = better (in PHP).

File: src/AbstractPathMappingRepository.php [hunk 13]
```diff
@@ -278,4 +300,49 @@ protected function createVirtualResource($path = null)
...
+            return $relativePath;
+        }
+
+        return Path::makeAbsolute($relativePath, $this->baseDirectory);
+    }
+
+    /**
+     * Transform a collection of relative paths into a collection of absolute paths.
+     *
+     * @param array $relativePaths
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: `string[]`

File: src/AbstractPathMappingRepository.php [hunk 14]
```diff
@@ -278,4 +300,49 @@ protected function createVirtualResource($path = null)
...
+     *
+     * @return array
```
Comment (webmozart) [inline src/AbstractPathMappingRepository.php]: `string[]`
